clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6183,7 +6203,98 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    // Generate forward declaration for free function.
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    if (K.SyclKernel->getLanguageLinkage() == CLanguageLinkage)
+      O << "extern \"C\" ";
+    FunctionTemplateDecl *FTD = K.SyclKernel->getPrimaryTemplate();
+    if (FTD)
+      FTD->getTemplateParameters()->print(O, S.getASTContext());
```

Comment:
Consider a free function:
```
template <int N, bool AIsVector, bool BIsVector>
SYCL_EXTERNAL SYCL_ESIMD_KERNEL SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::
         single_task_kernel)) void simd_f(uint32_t ValuesToTrySize,
                                          uint32_t *ValuesToTryPtr,
                                          uint32_t *BorrowMatrixPtr,
                                          uint32_t *ResultsMatrixPtr) {...
```}
We generate the following "shim" function for the free function, which gets used in a number of free function APIs:
```
template <int N, bool AIsVector, bool BIsVector> void simd_f(unsigned int, unsigned int *, unsigned int *, unsigned int *);
static constexpr auto __sycl_shim1() {
  return (void (*)(unsigned int, unsigned int *, unsigned int *, unsigned int *))simd_f<16, true, true>;
}
```
So, to answer your question, this is printed by lines 6219-20:
`template <int N, bool AIsVector, bool BIsVector>`

And to answer your second question, nothing currently prints the cast expression applied to simd_f that you see above in the body of the shim.

Reply:
>> So, to answer your first question, this is printed by lines 6219-20:
template <int N, bool AIsVector, bool BIsVector>

Please add a test which prints the template parameters then. Your current tests do not cover this, and I am not certain the printing will happen correctly with 'template' and opening and closing '<>'

>>And to answer your second question, nothing currently prints the cast expression applied to simd_f that you see above in the body of the shim. That's why it has to constructed by printing smaller parts.

I wasn't talking about printing the shim function. I was talking about printing the free function forward declaration. I understand my example incorrectly prints the body of the function as well, but there might be a way to avoid that by modifying the print functions/policy. If not, your code printing it by parts is ok. @AaronBallman are you aware of how we can print just the function forward declaration (no body) given a FunctionDecl?

Reply:
I found that printing the template declaration with Policy.SuppressDefinition suppresses the body, and then just a touchup of adding ";" creates a forward declaration.
Using this method the int header contains declarations that use template <......>. The int header test has been updated accordingly.